File: GamePad-Mobile/pc-server/Protocol/PacketDecoder.cs
namespace GamePadServer.Protocol;

/// <summary>
/// Decodes raw UDP data into typed Packet objects with CRC validation.
/// </summary>
public static class PacketDecoder
{
    /// <summary>
    /// Minimum valid packet size: 8 header + 2 CRC (no payload).
    /// </summary>
    private const int MinPacketSize = PacketHeader.ByteSize + 2;

    /// <summary>
    /// Decodes a byte array into a <see cref="Packet"/>.
    /// Returns null if the data is too short.
    /// </summary>
    public static Packet? Decode(byte[] data)
    {
        if (data.Length < MinPacketSize)
            return null;

        var header = PacketHeader.ReadFrom(data);
        var payloadLen = data.Length - PacketHeader.ByteSize - 2; // -2 for CRC
        var payload = new byte[payloadLen];

        if (payloadLen > 0)
            Buffer.BlockCopy(data, PacketHeader.ByteSize, payload, 0, payloadLen);

        var crcValid = Crc16.Validate(data);

        return new Packet
        {
            Header = header,
            Payload = payload,
            CrcValid = crcValid,
        };
    }
}

File: GamePad-Mobile/mobile/lib/core/protocol/packet_decoder.dart
import 'packet.dart';
import 'packet_header.dart';
import 'crc16.dart';

/// Decodes raw UDP bytes into a [Packet].
class PacketDecoder {
  static const int _minPacketSize = PacketHeader.byteSize + 2;

  /// Returns null if the data is too short or malformed.
  static Packet? decode(List<int> data) {
    if (data.length < _minPacketSize) return null;

    final header = PacketHeader.readFrom(data);
    final payloadLen = data.length - PacketHeader.byteSize - 2;
    final payload = data.sublist(PacketHeader.byteSize, PacketHeader.byteSize + payloadLen);

    final crcValid = Crc16.validate(data);

    return Packet(
      header: header,
      payload: payload,
      crcValid: crcValid,
    );
  }
}
